### PR TITLE
Add support for extended Unicode escape sequences in strings and templates

### LINKED ISSUE #2047: ES6 extended Unicode syntax for strings and template strings
`"\u{1f4a9}"` should be converted to `"\uD83D\uDCA9"`

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #2047.

ES6 extended escapes take the form of `\u{XXXXX...}` where `XXXXX...` is any number of hex digits. [It is a syntax error if the value of these digits is greater than 0x10FFFF](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-string-literals-static-semantics-early-errors). When interpreting the value of such a code point, it must be interpreted as two individual UTF16 codepoints through [a well-defined encoding scheme](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-utf16encoding).

Notable changes are:

* We support extended ES6 escape sequences as is.
* We support downlevel emit for ES6 escape sequences, using `\uXXXX` escapes for any non-ASCII characters in the original string.
* Template strings *always* use `\uXXXX` escapes for any non-ASCII characters.

Things we still don't support:

* Escapes in regular expressions.

### CODE REVIEW

File: src/compiler/diagnosticInformationMap.generated.ts [hunk 1]
```diff
@@ -155,6 +155,8 @@ module ts {
         Catch_clause_variable_name_must_be_an_identifier: { code: 1195, category: DiagnosticCategory.Error, key: "Catch clause variable name must be an identifier." },
         Catch_clause_variable_cannot_have_a_type_annotation: { code: 1196, category: DiagnosticCategory.Error, key: "Catch clause variable cannot have a type annotation." },
         Catch_clause_variable_cannot_have_an_initializer: { code: 1197, category: DiagnosticCategory.Error, key: "Catch clause variable cannot have an initializer." },
+        An_extended_Unicode_escape_value_must_be_between_0x0_and_0x10FFFF_inclusive: { code: 1198, category: DiagnosticCategory.Error, key: "An extended Unicode escape value must be between 0x0 and 0x10FFFF inclusive." },
+        expected: { code: 1199, category: DiagnosticCategory.Error, key: "'}' expected." },
```
Comment (CyrusNajmabadi) [inline src/compiler/diagnosticInformationMap.generated.ts]: 'Unterminated unicode escape'

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -611,7 +611,14 @@
         "category": "Error",
         "code": 1197
     },
-
+    "An extended Unicode escape value must be between 0x0 and 0x10FFFF inclusive.": {
+        "category": "Error",
+        "code": 1198
+    },
+    "'}' expected.": {
```
Comment (CyrusNajmabadi) [inline src/compiler/diagnosticMessages.json]: We already have "'{0}' expected."" for this.

File: src/compiler/scanner.ts [hunk 3]
```diff
@@ -764,11 +776,18 @@ module ts {
...
-                        return String.fromCharCode(ch);
+                    if (pos < len && text.charCodeAt(pos) === CharacterCodes.openBrace) {
+                        hasExtendedUnicodeEscape = true;
+                        pos++;
+                        return scanExtendedUnicodeEscape();
+                    }
+                    
+                    // fall through
+                case CharacterCodes.x:
+                    var escapedValue = scanExactNumberOfHexDigits(ch === CharacterCodes.x ? 2 : 4);
```
Comment (CyrusNajmabadi) [inline src/compiler/scanner.ts]: can you just extract this code, and then call it from the 'x' or 'u' cases?

File: src/compiler/scanner.ts [hunk 4]
```diff
@@ -790,6 +809,54 @@ module ts {
...
+        
+        // Derived from the 10.1.1 UTF16Encoding of the ES6 Spec.
+        function utf16EncodeAsString(codePoint: number): string {
+            Debug.assert(0x0 <= codePoint && codePoint <= 0x10FFFF);
+            
+            if (codePoint <= 65535) {
+                return String.fromCharCode(codePoint);
+            }
+            
+            var codeUnit1 = Math.floor((codePoint - 65536) / 1024) + 0xD800;
```
Comment (CyrusNajmabadi) [inline src/compiler/scanner.ts]: Can you just refernece the relevant portion of theES6 spec?

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -1130,7 +1130,7 @@ module ts {
             newEndN = Math.max(newEnd2, newEnd2 + (newEnd1 - oldEnd2));
         }
 
-        return createTextChangeRange(createTextSpanFromBounds(oldStartN, oldEndN), /*newLength: */newEndN - oldStartN);
+        return createTextChangeRange(createTextSpanFromBounds(oldStartN, oldEndN), /*newLength: */ newEndN - oldStartN);
```
Comment (CyrusNajmabadi) [inline src/compiler/utilities.ts]: remove space after :

File: src/compiler/utilities.ts [hunk 6]
```diff
@@ -1212,4 +1212,53 @@ module ts {
             }
         }
     }
+    
+    var backslashOrDoubleQuote = /[\"\\]/g;
+    var escapedCharsRegExp = /[\u0000-\u001f\t\v\f\b\r\n\u2028\u2029\u0085]/g;
```
Comment (CyrusNajmabadi) [inline src/compiler/utilities.ts]: Document this.

File: src/compiler/scanner.ts (hunk 4 above)
Comment (DanielRosenwasser) [inline src/compiler/scanner.ts]: I did reference the relevant portion of the ES6 spec, it's at the top of the function.

Comment (DanielRosenwasser) [inline src/compiler/scanner.ts]: It's exactly the same as 10.1.1

**Pull request merged**